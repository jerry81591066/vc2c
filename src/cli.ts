import program from 'commander'
import { convertFile } from './index.js'
import inquirer from 'inquirer'
import { writeFileInfo } from './file'

program
  .version('0.0.1')
  .usage('<command> [options]')

program
  .command('single <filePath>')
  .description('convert vue component file from class to composition api')
  .option('-v, --view', 'output file content on stdout, and no write file.')
  .option('-o, --output', 'output result file path')
  .option('-r, --root <root>', 'set root path for calc file absolute path, default: `process.cwd()`')
  .option('-c, --config <config>', 'set vc2c config file path, default: `\'.vc2c.js\'`')
  // eslint-disable-next-line @typescript-eslint/no-misused-promises
  .action(async (filePath: string, cmd) => {
    const cmdOptions = getCmdOptions(cmd)
    if (!cmdOptions.output && !cmdOptions.view) {
      const result = await inquirer.prompt({
        name: 'ok',
        type: 'confirm',
        message: 'You aren\'t using -o option to set output file path, It will replace original file content.'
      })
      if (!result.ok) {
        return
      }
    }

    const { file, result } = convertFile(filePath, cmdOptions.root, cmdOptions.config)
    if (cmdOptions.view) {
      // eslint-disable-next-line no-console
      console.log(result)
      return
    }

    writeFileInfo(file, result)
  })

program.parse(process.argv)

function camelize (str: string) {
  return str.replace(/-(\w)/g, (_, c) => c ? c.toUpperCase() : '')
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
function getCmdOptions (cmd: any) {
  const args: { [key: string]: boolean | string } = {}
  cmd.options.forEach((o: { long: string }) => {
    const key = camelize(o.long.replace(/^--/, ''))

    if (typeof cmd[key] !== 'function' && typeof cmd[key] !== 'undefined') {
      args[key] = cmd[key]
    }
  })
  return args
}
