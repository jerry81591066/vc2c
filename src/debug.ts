let debugMode = false

export function setDebugMode (value: boolean) {
  debugMode = value
}

export function log (message: string) {
  if (debugMode) {
    // eslint-disable-next-line no-console
    console.log(`${new Date().toLocaleString()}: ${message}`)
  }
}
