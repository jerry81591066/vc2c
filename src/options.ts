import * as ts from 'typescript'
import { ASTConvertPlugins } from './plugins/types'
import { getDefaultPlugins } from './plugins'
import * as vueTemplateParser from 'vue-template-compiler'

export interface Vc2cOptions {
  root: string
  debug: boolean,
  compatible: boolean
  setupPropsKey: string
  setupContextKey: string
  typesciprt: typeof ts
  vueTemplateParser: typeof vueTemplateParser
  eslintConfigFile: string
  plugins: ASTConvertPlugins
}

export type InputVc2cOptions = Partial<Vc2cOptions>

export function getDefaultVc2cOptions (tsModule: typeof ts = ts): Vc2cOptions {
  return {
    root: process.cwd(),
    debug: false,
    compatible: false,
    setupPropsKey: 'props',
    setupContextKey: 'context',
    typesciprt: tsModule,
    vueTemplateParser: vueTemplateParser,
    eslintConfigFile: '.eslintrc.js',
    plugins: getDefaultPlugins(tsModule)
  }
}

export function mergeVc2cOptions (original: Vc2cOptions, merged: InputVc2cOptions): Vc2cOptions {
  return {
    ...original,
    ...merged
  }
}
