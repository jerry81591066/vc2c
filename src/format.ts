import { Vc2cOptions } from './options'
import path from 'path'
import fs from 'fs'
import prettierFormat from 'prettier-eslint'
import { log } from './debug'

export function format (content: string, options: Vc2cOptions) {
  const eslintConfigPath = path.resolve(options.root, options.eslintConfigFile)
  const prettierEslintOpions = (fs.existsSync(eslintConfigPath))
    ? {
      text: content,
      filePath: eslintConfigPath,
      prettierOptions: {
        parser: 'typescript'
      },
      fallbackPrettierOptions: {
        parser: 'typescript'
      }
    }
    : {
      text: content,
      filePath: '',
      eslintConfig: {
        parser: require.resolve('@typescript-eslint/parser'),
        parserOptions: {
          sourceType: 'module',
          ecmaFeatures: {
            jsx: false
          }
        },
        rules: {
          semi: ['error', 'never'],
          'padding-line-between-statements': [
            'error',
            { blankLine: 'always', prev: '*', next: 'export' },
            { blankLine: 'always', prev: 'const', next: 'const' },
            { blankLine: 'always', prev: 'const', next: 'return' }
          ]
        }
      },
      prettierOptions: {
        parser: 'typescript',
        Semicolons: false,
        singleQuote: true
      },
      fallbackPrettierOptions: {
        parser: 'typescript',
        singleQuote: true,
        Semicolons: false
      }
    }

  log('Format result code.....')
  return prettierFormat(prettierEslintOpions)
}
