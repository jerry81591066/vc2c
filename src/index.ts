import { getSingleFileProgram } from './parser'
import { convertAST } from './convert'
import { InputVc2cOptions, getDefaultVc2cOptions, mergeVc2cOptions } from './options'
import { format } from './format'
import path from 'path'
import fs from 'fs'
import { readVueSFCOrTsFile } from './file'
import { setDebugMode } from './debug'
import * as BuiltInPlugins from './plugins/builtIn'

export function convert (content: string, inputOptions: InputVc2cOptions) {
  const options = mergeVc2cOptions(getDefaultVc2cOptions(inputOptions.typesciprt), inputOptions)
  const { ast, program } = getSingleFileProgram(content, options)

  return format(convertAST(ast, options, program), options)
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export function convertFile (filePath: string, root: any, config: any) {
  root = (typeof root === 'string')
    ? (
      path.isAbsolute(root) ? root : path.resolve(process.cwd(), root)
    )
    : process.cwd()
  config = (typeof config === 'string') ? config : '.vc2c.js'
  const inputOptions: InputVc2cOptions = fs.existsSync(path.resolve(root, config))
    ? require(path.resolve(root, config))
    : {}
  const options = mergeVc2cOptions(getDefaultVc2cOptions(inputOptions.typesciprt), inputOptions)
  options.root = root

  if (options.debug) {
    setDebugMode(true)
  }

  const file = readVueSFCOrTsFile(filePath, options)
  return {
    file,
    result: convert(file.content, options)
  }
}

export * from './plugins/types'
export { BuiltInPlugins }
export * from './utils'
