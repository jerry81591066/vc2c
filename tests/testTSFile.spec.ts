import { convertFile } from '../src'
import { FileKind } from '../src/file'
import path from 'path'

describe('testTSFile', () => {
  const filePath = 'fixture/Input.ts'

  it('compatible', () => {
    const { file, result } = convertFile(filePath, __dirname, 'config/.compatible.vc2c.js')
    expect(file.fsPath.includes(filePath)).toBeTruthy()
    expect(path.isAbsolute(file.fsPath)).toBeTruthy()
    expect(file.kind).toBe(FileKind.TS)
    expect(file).not.toHaveProperty('start')
    expect(file).not.toHaveProperty('end')
    expect(result).toMatchSnapshot()
  })

  it('no compatible', () => {
    const { file, result } = convertFile(filePath, __dirname, 'config/.nocompatible.vc2c.js')
    expect(file.fsPath.includes(filePath)).toBeTruthy()
    expect(path.isAbsolute(file.fsPath)).toBeTruthy()
    expect(file.kind).toBe(FileKind.TS)
    expect(file).not.toHaveProperty('start')
    expect(file).not.toHaveProperty('end')
    expect(result).toMatchSnapshot()
  })
})
