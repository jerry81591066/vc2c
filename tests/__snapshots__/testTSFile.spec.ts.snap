// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testTSFile compatible 1`] = `
"import {
  computed,
  reactive,
  toRefs,
  ref,
  watch,
  createComponent
} from '@vue/composition-api'

/**
 * My basic tag
 */
export default createComponent({
  model: {
    prop: 'checked',
    event: 'change'
  },
  props: {
    bar: String,
    checked: { type: Boolean },
    /**
     * My foo
     */
    foo: { type: Number, default: false }
  },
  name: 'oao',
  setup(props, context) {
    /**
     * My count
     */
    const count = computed(() => {
      return context.$store.state.count
    })

    const { a } = (() => {
      const a = 'pa'

      return toRefs(
        reactive({
          a: a
        })
      )
    })()

    const anotherComponent = ref<HTMLElement>(null)

    /**
     * My msg
     */
    const msg = ref('Vetur means \\"Winter\\" in icelandic.') // foo
    watch(
      checked,
      (val: boolean, newVal: boolean) => {
        console.log(val, newVal)
      },
      { deep: true, lazy: false }
    )
    const onMsgChanged = (val: string, newVal: string) => {
      console.log(val, newVal)
    }

    const click = (c: string) => {
      context.$emit('click', c)
    }

    const removeItem = () => {
      if (props.foo > 10) {
        context.$emit('remove-item', 10)
      } else {
        context.$emit('remove-item', props.foo)
      }
    }

    /**
     * My greeting
     */
    const hello = () => {
      console.log(msg.value)
    }

    // TODO: Please remove unused return variable
    return {
      count,
      a,
      anotherComponent,
      msg,
      checked,
      onMsgChanged,
      click,
      remove,
      hello,
      bar: props.bar,
      checked: props.checked,
      foo: props.foo
    }
  }
})
"
`;

exports[`testTSFile no compatible 1`] = `
"import { computed, reactive, toRefs, ref, watch } from 'vue'

/**
 * My basic tag
 */
export default {
  model: {
    prop: 'checked',
    event: 'change'
  },
  props: {
    bar: String,
    checked: { type: Boolean },
    /**
     * My foo
     */
    foo: { type: Number, default: false }
  },
  name: 'oao',
  setup(props, context) {
    /**
     * My count
     */
    const count = computed(() => {
      return context.$store.state.count
    })

    const { a } = (() => {
      const a = 'pa'

      return toRefs(
        reactive({
          a: a
        })
      )
    })()

    const anotherComponent = ref<HTMLElement>(null)

    /**
     * My msg
     */
    const msg = ref('Vetur means \\"Winter\\" in icelandic.') // foo
    watch(
      checked,
      (val: boolean, newVal: boolean) => {
        console.log(val, newVal)
      },
      { deep: true, lazy: false }
    )
    const onMsgChanged = (val: string, newVal: string) => {
      console.log(val, newVal)
    }

    const click = (c: string) => {
      context.emit('click', c)
    }

    const removeItem = () => {
      if (props.foo > 10) {
        context.emit('remove-item', 10)
      } else {
        context.emit('remove-item', props.foo)
      }
    }

    /**
     * My greeting
     */
    const hello = () => {
      console.log(msg.value)
    }

    // TODO: Please remove unused return variable
    return {
      count,
      a,
      anotherComponent,
      msg,
      checked,
      onMsgChanged,
      click,
      remove,
      hello,
      bar: props.bar,
      checked: props.checked,
      foo: props.foo
    }
  }
}
"
`;
